### ISSUE TITLE
[sival,test] Flakiness in detecting PASS! in Rust test harnesses

### ISSUE THREAD

**Issue by jwnrt** [lowRISC/opentitan#20580]
### Description

The `pwrmgr_random_sleep_all_wake_ups_fpga_cw310_test_rom` has been seen to fail on some runs, possibly due to a race / bug in the host side of the test.

<details><summary> Here's the Azure logs for a run where this test failed (expand):</summary>

```
==================== Test output for //sw/device/tests:pwrmgr_random_sleep_all_wake_ups_fpga_cw310_test_rom:
Invoking test: sw/host/tests/chip/pwrmgr/sleep_all_wakeups --rcfile= --logging=info --interface=cw310 --uarts=/dev/ttyACM_CW310_1,/dev/ttyACM_CW310_0 --bootstrap=sw/device/tests/pwrmgr_random_sleep_all_wake_ups_fpga_cw310_test_rom.bin --bitstream=hw/bitstream/universal/splice.bit
[2023-12-06T13:07:10.954Z INFO  opentitanlib::util::usb] Could not open device at bus=1 address=7: No such device (it may have been disconnected)
[2023-12-06T13:07:10.954Z INFO  opentitanlib::util::usb] Could not open device at bus=1 address=100: No such device (it may have been disconnected)
[2023-12-06T13:07:10.954Z INFO  opentitanlib::util::usb] Could not open device at bus=1 address=118: No such device (it may have been disconnected)
[2023-12-06T13:07:10.954Z INFO  opentitanlib::util::usb] Could not open device at bus=1 address=109: No such device (it may have been disconnected)
[2023-12-06T13:07:10.954Z INFO  opentitanlib::util::usb] Could not open device at bus=1 address=35: No such device (it may have been disconnected)
[2023-12-06T13:07:10.959Z INFO  opentitanlib::io::uart] set_flow_control to false
[2023-12-06T13:07:10.959Z INFO  opentitanlib::test_utils::load_bitstream] Loading bitstream: "hw/bitstream/universal/splice.bit"
[2023-12-06T13:07:10.972Z INFO  opentitanlib::util::usr_access] Bitstream file USR_ACCESS value: 0xc6bcc37b
[2023-12-06T13:07:11.035Z INFO  opentitanlib::util::rom_detect] Current bitstream: "TestROM:c6bcc37b\r"
[2023-12-06T13:07:11.035Z INFO  opentitanlib::transport::common::fpga] Already running the correct bitstream.  Skip loading bitstream.
[2023-12-06T13:07:11.036Z INFO  opentitanlib::test_utils::init] load_bitstream: success.
"load_bitstream": true
[2023-12-06T13:07:11.037Z INFO  opentitanlib::bootstrap] Asserting bootstrap pins...
[2023-12-06T13:07:11.037Z INFO  opentitanlib::app] Asserting the reset signal
[2023-12-06T13:07:11.138Z INFO  opentitanlib::app] Clearing the UART RX buffer
[2023-12-06T13:07:11.138Z INFO  opentitanlib::app] Deasserting the reset signal
[2023-12-06T13:07:11.238Z INFO  opentitanlib::bootstrap] Performing bootstrap...
[2023-12-06T13:07:14.099Z INFO  opentitanlib::bootstrap] Releasing bootstrap pins, resetting device...
[2023-12-06T13:07:14.101Z INFO  opentitanlib::app] Asserting the reset signal
[2023-12-06T13:07:14.201Z INFO  opentitanlib::app] Deasserting the reset signal
[2023-12-06T13:07:14.301Z INFO  opentitanlib::test_utils::init] bootstrap: success.
"bootstrap": true
68:Starting test sleep_all_wakeups_test...
[2023-12-06T13:07:14.302Z  console]I00001 test_rom.c:158] kChipInfo: scm_revision=e0024355
[2023-12-06T13:07:14.302Z  console]I00002 test_rom.c:184] TestROM:c6bcc37b
[2023-12-06T13:07:14.302Z  console]I00003 test_rom.c:190] Boot strap requested
[2023-12-06T13:07:14.302Z  console]I00001 test_rom.c:158] kChipInfo: scm_revision=e0024355
[2023-12-06T13:07:14.302Z  console]I00002 test_rom.c:184] TestROM:c6bcc37b
[2023-12-06T13:07:14.302Z  console]I00003 test_rom.c:230] Test ROM complete, jumping to flash (addr: 20000480)!
[2023-12-06T13:07:14.302Z  console]I00001 ottf_main.c:154] Running sw/device/tests/pwrmgr_random_sleep_all_wake_ups.c
[2023-12-06T13:07:14.302Z  console]I00002 pwrmgr_sleep_all_wake_ups_impl.c:225] Issue WFI to enter sleep 0

[2023-12-06T13:07:14.302Z  console]I00003 pwrmgr_random_sleep_all_wake_ups.c:121] Woke up by source 0
[2023-12-06T13:07:14.309Z  console]
[2023-12-06T13:07:14.309Z  console]I00004 pwrmgr_sleep_all_wake_ups_impl.c:225] Issue WFI to enter sleep 0
[2023-12-06T13:07:14.315Z  console]
[2023-12-06T13:07:14.319Z  console]I00001 test_rom.c:158] kChipInfo: scm_revision=e0024355
[2023-12-06T13:07:14.324Z  console]I00002 test_rom.c:184] TestROM:c6bcc37b
[2023-12-06T13:07:14.328Z  console]I00003 test_rom.c:230] Test ROM complete, jumping to flash (addr: 20000480)!
[2023-12-06T13:07:14.335Z  console]I00001 ottf_main.c:154] Running sw/device/tests/pwrmgr_random_sleep_all_wake_ups.c
[2023-12-06T13:07:14.343Z  console]I00002 pwrmgr_random_sleep_all_wake_ups.c:94] Woke up by source 0
[2023-12-06T13:07:14.349Z  console]
[2023-12-06T13:07:14.349Z  console]I00003 pwrmgr_sleep_all_wake_ups_impl.c:225] Issue WFI to enter sleep 2
[2023-12-06T13:07:14.355Z  console]
[2023-12-06T13:07:14.356Z  console]I00004 pwrmgr_random_sleep_all_wake_ups.c:121] Woke up by source 2
[2023-12-06T13:07:14.362Z  console]
[2023-12-06T13:07:14.362Z  console]I00005 pwrmgr_sleep_all_wake_ups_impl.c:225] Issue WFI to enter sleep 2
[2023-12-06T13:07:14.369Z  console]
[2023-12-06T13:07:14.372Z  console]I00001 test_rom.c:158] kChipInfo: scm_revision=e0024355
[2023-12-06T13:07:14.378Z  console]I00002 test_rom.c:184] TestROM:c6bcc37b
[2023-12-06T13:07:14.381Z  console]I00003 test_rom.c:230] Test ROM complete, jumping to flash (addr: 20000480)!
[2023-12-06T13:07:14.388Z  console]I00001 ottf_main.c:154] Running sw/device/tests/pwrmgr_random_sleep_all_wake_ups.c
[2023-12-06T13:07:14.396Z  console]I00002 pwrmgr_random_sleep_all_wake_ups.c:94] Woke up by source 2
[2023-12-06T13:07:14.402Z  console]
[2023-12-06T13:07:14.402Z  console]I00003 pwrmgr_sleep_all_wake_ups_impl.c:131] prgm_usb_wakeup: wait 20us (usb)
[2023-12-06T13:07:14.409Z  console]I00004 pwrmgr_sleep_all_wake_ups_impl.c:225] Issue WFI to enter sleep 3
[2023-12-06T13:07:14.416Z  console]
[2023-12-06T13:07:14.416Z  console]I00005 pwrmgr_random_sleep_all_wake_ups.c:121] Woke up by source 3
[2023-12-06T13:07:14.422Z  console]
[2023-12-06T13:07:14.422Z  console]I00006 pwrmgr_sleep_all_wake_ups_impl.c:131] prgm_usb_wakeup: wait 20us (usb)
[2023-12-06T13:07:14.429Z  console]I00007 pwrmgr_sleep_all_wake_ups_impl.c:225] Issue WFI to enter sleep 3
[2023-12-06T13:07:14.436Z  console]
[2023-12-06T13:07:14.440Z  console]I00001 test_rom.c:158] kChipInfo: scm_revision=e0024355
[2023-12-06T13:07:14.445Z  console]I00002 test_rom.c:184] TestROM:c6bcc37b
[2023-12-06T13:07:14.448Z  console]I00003 test_rom.c:230] Test ROM complete, jumping to flash (addr: 20000480)!
[2023-12-06T13:07:14.456Z  console]I00001 ottf_main.c:154] Running sw/device/tests/pwrmgr_random_sleep_all_wake_ups.c
[2023-12-06T13:07:14.463Z  console]I00002 pwrmgr_random_sleep_all_wake_ups.c:94] Woke up by source 3

[2023-12-06T13:07:14.470Z  console]I00003 pwrmgr_sleep_all_wake_ups_impl.c:225] Issue WFI to enter sleep 4
[2023-12-06T13:07:15.053Z  console]I00004 pwrmgr_random_sleep_all_wake_ups.c:121] Woke up by source 4
[2023-12-06T13:07:15.053Z  console]I00005 pwrmgr_sleep_all_wake_ups_impl.c:225] Issue WFI to enter sleep 4
[2023-12-06T13:07:15.053Z  console]I00001 test_rom.c:158] kChipInfo: scm_revision=e0024355
[2023-12-06T13:07:15.053Z  console]I00002 test_rom.c
[2023-12-06T13:07:15.053Z  console]:184] TestROM:c6bcc37b
[2023-12-06T13:07:15.053Z  console]I00003 test_rom.c:230] Test ROM complete, jumping to flash (addr: 20000480)!
[2023-12-06T13:07:15.053Z  console]I00001 ottf_main.c:154] Running sw/device/tests/pwrmgr_random_sleep_all_wake_ups.c
[2023-12-06T13:07:15.053Z  console]I00002 pwrmgr_random_sleep_all_wake_ups.c:94] Woke up by source 4
[2023-12-06T13:07:15.053Z  console]I00
[2023-12-06T13:07:15.053Z  console]003 pwrmgr_sleep_all_wake_ups_impl.c:225] Issue WFI to enter sleep 5
[2023-12-06T13:07:15.053Z  console]I00004 pwrmgr_random_sleep_all_wake_ups.c:121] Woke up by source 5
[2023-12-06T13:07:15.053Z  console]I00005 pwrmgr_sleep_all_wake_ups_impl.c:225] Issue WFI to enter sleep 5
[2023-12-06T13:07:15.053Z  console]I00001 test_rom.c:158] kChipInfo: scm_revisio
[2023-12-06T13:07:15.053Z  console]n=e0024355
[2023-12-06T13:07:15.053Z  console]I00002 test_rom.c:184] TestROM:c6bcc37b
[2023-12-06T13:07:15.053Z  console]I00003 test_rom.c:230] Test ROM complete, jumping to flash (addr: 20000480)!
[2023-12-06T13:07:15.053Z  console]I00001 ottf_main.c:154] Running sw/device/tests/pwrmgr_random_sleep_all_wake_ups.c
[2023-12-06T13:07:15.053Z  console]I00002 pwrmgr_random_sleep_all_wake_ups.c:94] Woke up by source 5
[2023-12-06T13:07:15.053Z  console]I00003 ottf_main.c:109] Finished sw/device/tests/pwrmgr_random_sleep_all_wake_ups.c
[2023-12-06T13:07:15.053Z  console]I00004 status.c:28] PASS!

-- Test timed out at 2023-12-06 13:08:10 UTC --
================================================================================

```

</details>

These logs show that the test did print the `PASS!` message, but still timed out:

```
[2023-12-06T13:07:15.053Z  console]I00002 pwrmgr_random_sleep_all_wake_ups.c:94] Woke up by source 5
[2023-12-06T13:07:15.053Z  console]I00003 ottf_main.c:109] Finished sw/device/tests/pwrmgr_random_sleep_all_wake_ups.c
[2023-12-06T13:07:15.053Z  console]I00004 status.c:28] PASS!
```

I'm not really sure why, but perhaps some race caused the `PASS!` line to be absorbed by the check for `Woke up by source...` further down?:

https://github.com/lowRISC/opentitan/blob/c009a4a71127b82106de9faf18ab96fd809936fa/sw/host/tests/chip/pwrmgr/src/sleep_all_wakeups.rs#L52

**Comment by pamaury** [comment]
I think I already raised this issue in one PR of mattute but I can't find it again. I agree that the behaviour of `wait_for` is super counterintuitive. The proble, as far as I can tell, is here: in [`uart_read`](https://github.com/lowRISC/opentitan/blob/master/sw/host/opentitanlib/src/uart/console.rs#L179). It reads a full buffer of characters (256 bytes) but if there is a match mid-buffer, it throws away the rest of it. Some solutions could be:
- read one character at a time: might be inefficient
- read a full buffer but `put`-back the rest of it in the Uart if there is a match. This might require to modify `ConsoleDevice` to add support for that.

**Comment by pamaury** [comment]
I think it makes sense but the way the `ConsoleDevice` is implemented is quite convoluted, I honestly have no idea where to start.

**Comment by a-will** [comment]
It's goofy because `UartConsole::wait_for()` creates a new UartConsole on every entry and drops it upon return, and that object is the final holder of all the data it greedily scooped up from the UART (held in the `buffer` String member).

This seems like a peculiar ownership model, since any overzealous function call could discard data that is relevant to another. Would it make sense to have a more explicit single-producer/multi-consumer stream model? Register/unregister listeners, copy streams (or manage different sets of indices and/or reference counts, I guess), and have each listener manage their own stream?

Right now, the stdout and logging outputs are kind of mashed together with other pattern matching calls in UartConsole. It seems to have grown into something that is difficult to maintain.

**Comment by matutem** [comment]
I raised the priority and added SiliconValidation since this causes spurious failures and will slow us down.

**Comment by jesultra** [comment]
I see two possible solutions.

1. We could let `UartConsole` keep a buffer of the characters it has read beyond the last match in `wait_for()`, that would require making `wait_for()` taking a `&self` parameter as well as changing e.g. `exit_success` from an instance variable to a parameter into `wait_for()`, and it would require all callers to pass along the `UartConsole` instance, rather than getting the lower level `Uart` object and each creating their own `UartConsole`.  I do not think this is a bad design at all, but it could require changing lots of places.
2. We could implement an interface to "put back" characters to a `Uart` object.  As has been noted, there are currently several implementations of the `Uart` trait, and I think it is undesirable to put the burden on each one of them (and on future implementation) to support putting back characters.  Rather, I would suggest that we put the functionality at the `TranportWrapper` layer.  That is, we would add a new `put_back()` method to the `Uart` trait, with the undestanding that none of the backends will implement that method.  Then we would change [TransportWrapper::uart()](https://github.com/lowRISC/opentitan/blob/38d5796f7f93f5916dca663c6010fc6483f279bb/sw/host/opentitanlib/src/app/mod.rs#L783) such that instead of returning the underlying `Uart` object directly, it would wrap it in another object which implements the `Uart` trait, and which has a buffer of characters put back.  The `read()` method on this wrapper object would inspect the put-back buffer, and return immediately if present, or call through to the underlying `Uart` object only if empty.  This will be a relatively narrow change, and would align with how I2C and SPI objects are also wrapped at this layer.  One possible complication is `register_nonblocking_read()` and related functions, which allows callers to set up the Uart for use by `mio::Poll`, I think this can be worked around, by always reading from the Uart at the top of `interact_mio`, though it is ugly to have two partially incompatible interfaces in the same `Trait`.  Maybe the first proposal is the better.

**Comment by timothytrippel** [comment]
Am I missing something, or will simply reading 1 char at a time (instead of 256), as implemented in #20592, not fix this problem? Sure it will hurt performance a bit, but the UART is already super slow compared to host processing speed. Will it really matter that much?

Longer term we could think about more performant alternatives, but it seems there are other higher priority issues to address ATM, if simply reading one char at a time will resolve this now.

**Comment by jesultra** [comment]
Tim, you are right.  Reading one character at a time solves the issue without the drawbacks listed in my two proposals above.  That should probably be what we try first.
